{% extends "unitbase.html" %}
{% load pytags %}
{# title = Filtering SAM #}
{# subtitle = selecting alignments #}
{# name = Filtering SAM files #}

{% block body %}

{% md sam %}
    
> **Reference:** The [SAM format specification][samspec]. Keep it under your pillow.

[samspec]: https://samtools.github.io/hts-specs/

- - -
    
Prepare the genome. We'll be using the Ebola genome of the 1976 outbreak.

    # Set up the various shortcuts
    alias bam='samtools view -b - | samtools sort'
    REF=~/refs/ebola/1976.fa
    R1=SRR1972739_1.fastq
    R2=SRR1972739_2.fastq

    # Obtain and prepare the genome.
    mkdir -p ~/refs/ebola
    efetch -db=nuccore -format=fasta -id=AF086833 > $REF
    bwa index $REF

    # Obtain the data
    fastq-dump -X 1000 --split-files SRR1972739

    # Align the reads.
    bwa mem $REF $R1 $R2 | bam > bwa.bam

    # Create the index
    samtools index bwa.bam

    # You can also convert flags at the command line.
    samtools flags

    # This will tell you the number
    samtools flags REVERSE
    # 0x10	16	REVERSE

    # Mapping quality over 1. For BWA that means uniquely mapping.
    samtools view -c -q 1 bwa.bam
    # 1549

    # Mapping quality over 1 and on the reverse strand.
    samtools view -c -q 1 -F 4 -f 16  bwa.bam
    # 651

    # Mapping quality over 1 and on the forward strand.
    samtools view -c -q 1 -F 4 -F 16 bwa.bam
    # 898

    # We can build complex statements via the flags:
    # Mapping quality over 1 + read is paired
    # read maps to reverse strand + read mate is also on reverse strand.
    #
    # This is a weird case actually. Could be either be an interesting
    # case or just some systematic sequencing error ... usually is the latter.
    samtools flags PAIRED,REVERSE,MREVERSE
    # 0x31	49	PAIRED,REVERSE,MREVERSE

    samtools view -c -q 1 -F 4 -f 51 bwa.bam
    # 14

    # Compute the depth of coverage.
    samtools depth bwa.bam  | head

    # We could sort by depth though that
    # (depending on genome size could take a long time)
    # Find the highest covered region
    samtools depth bwa.bam  | sort -k 3 -rn | head

    # Other reporting commands
    samtools flagstat bwa.bam

    # Reference sequence name, sequence length, number of mapped reads and # unmapped reads.
    # How can a read be "unmapped" to a reference?
    # https://www.biostars.org/p/14569/#14570
    samtools idxstats bwa.bam

{% endmd %}

{% endblock %}
