{% extends "unitbase.html" %}
{% load pytags %}
{# title = SAM #}
{# subtitle = sequence alignment maps #}
{# name = Representing billions of alignments: The Sequence Alignment Maps (SAM) #}

{% block body %}

{% md sam %}

A SAM file contains so called information on alignments.

> **Reference:** The [SAM format specification][sam-spec]. Keep it under your pillow.

- - -

### Creating SAM files

Please refer to {%  link 'short-read-aligners' %} on how to generate these.
This tutorial will assume that you have at least the `bwa.sam`  file
generated with instruction on that page.

### Sam Flags

The samtools `flags` allow us to select alignments that have certain properties.
As of samtools 1.2 there is a `flags` command to covert the integer codes to human readable output:

    # List all flags.
    samtools flags

    # Combine flags into a single number.
    samtools flags PAIRED,PROPER_PAIR,REVERSE,MREVERSE
    # 0x33    51    PAIRED,PROPER_PAIR,REVERSE,MREVERSE

    # It works backwards as well.
    samtools flags 51
    # 0x33	51	PAIRED,PROPER_PAIR,REVERSE,MREVERSE

[explain-sam]: https://broadinstitute.github.io/picard/explain-flags.html
[sam-spec]: https://samtools.github.io/hts-specs/

- - -

{%  endmd %}

{% md format 'Explore the SAM format' %}

Prepare a genome for alignment. We're using the Ebola genome of the 1976 outbreak.

    mkdir -p ~/refs/ebola
    REF=~/refs/ebola/1976.fa
    efetch -db=nuccore -format=fasta -id=AF086833 > $REF
    bwa index $REF

What we want to prepare are four sequences with known properties and then we will align each
separately:

* A sequence that is an exact match,
* A sequence from the reverse strand,
* A sequence with a mismatch and,
* A sequence with an insertion or deletion

We will use the emboss tools called [seqret](http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/seqret.html)
and [msbar](http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/msbar.html), powerful
but somewhat unfriendly programs.
To mutate the sequence you could edit it in an editor or use the `msbar` EMBOSS tool.
It has the following values to its `-point` parameter.

    -point  Types of point mutations to perform
             0 None
             1 Any of the following
             2 Insertions
             3 Deletions
             4 Changes
             5 Duplications
             6 Moves

This will create a file called `query.fa` with mutated sequences.

    # Extract the first 50 bases, rename the sequence 'exact'
    seqret -filter -send 50  -sid exact $REF > query.fa

    # Extract the first 50 bases, reverse complement it, rename the sequence 'reverse'
    seqret -filter -send 50 -srev -sid reverse $REF >> query.fa

    # Mutate with a mismatch. Rename the sequence to `mismatch`
    seqret -filter -send 50 $REF | msbar -filter -sid mismatch -point 4 >> query.fa

    # Mutate with an insertion. Rename the sequence to `insertion`
    seqret -filter -send 50 $REF | msbar -filter -sid insertion -point 2 >> query.fa

To visualize the alignments we could run `lastz` with so called text output

    lastz $REF query.fa --format=text

There will be an alignment for each query sequence. For example in my case
the last one was. It shows the insertion of the bottom query sequence
relative to the reference (top) sequence:

    1 CGGACACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGT-GCGAAT
      |||||||||||||||||||||||||||||||||||||||||||-||||||
    1 CGGACACACAAAAAGAAAGAAGAATTTTTAGGATCTTTTGTGTGGCGAAT

Align the same with bwa and let's look at what the SAM file structure
is like:

    bwa mem $REF query.fa > query.sam

From this SAM file cut different columns and look at their content:

    # First three columns: QNAME, FLAG, RNAME
    cat query.sam | cut -f 1,2,3

    # Second three columns: POS, MAPQ, CIGAR
    cat query.sam | cut -f 4,5,6

    # Third three columns: RNEXT, PNEXT, TLEN
    # These are set only for paired end sequencing runs.
    cat query.sam | cut -f 7,8,9

    # Last two columns: SEQ, QUAL
    # No qualities in this case since it was a FASTA file.
    cat query.sam | cut -f 10,11

    # The remaining columns are optional. Some of the most commonly filled ones:
    # NM: number of mismatches
    # AS: Alignment score
    # MD: Mismatching positions
    # NH: Number of reported alignments that contains the query
    # Read the SAM spec for the definition of TAG names that don't start with X.
    # Read the aligner manual for the definition of TAG names that start with X
    cat query.sam | cut -f 12,13,14

{% endmd %}

{% md samtools-basics 'Using Samtools: Basics' %}

Samtools is a magical tool that alone could be used
to reproduce the majority of results in a suprising number publications.

{% code "sam/samtools-commands.sh" %}

{% endmd %}

{% md advanced 'Using Samtools: Advanced usage' %}

Evaluate simulated datasets

{% code "sam/samtools-wgsim.sh" %}
{% endmd %}

{% md variation 'Using Samtools: Large scale rearrangments, pileups' %}

Use samtools to investigate genome structure. Generate pileups.

{% code "sam/samtools-variation.sh" %}
{% endmd %}

{% endblock %}
