{% extends "unitbase.html" %}
{% load pytags %}
{# title = SAM #}
{# subtitle = sequence alignment maps #}
{# name = Alignment representation: Sequence Alignment Maps (SAM) #}

{% block body %}

{% markdown %}

This page contains the following sections:

* [Explore the SAM Format](#format).
  <br>You understand the SAM format when
  you are able to describe how a change in a genome will show up
  in the SAM file.

* [Using Samtools: Basics](#samtools-basics)
  <br>Converting to `BAM`, filtering alignments, using commands:
  `view`, `depth`, `tview`

* [Using Samtools: Advanced usage](#advanced)
  <br>Evaluate simulated datasets, readgroups, processing

* [Using Samtools: Pileups](#pileups)
  <br>Samtool pileup, a poor man's SNP caller.

Software installation help: {% link "install-everything" %}

- - -

The [SAM format][sam-spec]
is so easy to use that we needed a dedicated website: [Explain SAM Flags][explain-sam]
to decipher a single field of it.
Good news though, as of samtools 1.2 there is a `flags` command to
covert the integer codes to human readable output:

{% code "sam/samtools-flags.sh" %}

[explain-sam]: https://broadinstitute.github.io/picard/explain-flags.html
[sam-spec]: https://samtools.github.io/hts-specs/
- - -
{%  endmarkdown %}

{% md format %}
### Explore the SAM format

The Sequence Alignment Map definition: [SAM Specification][sam-spec]

[sam-spec]: https://samtools.github.io/hts-specs/

{% code "sam/explore-sam-format.sh" %}

{% endmd %}

{% md samtools-basics %}
### Using Samtools: Basics

Samtools is a magical tool that alone could be used
to reproduce the majority of results in a suprising number publications.

{% code "sam/samtools-commands.sh" %}
{% endmd %}

{% md advanced %}
### Using Samtools: Advanced usage

Evaluate simulated datasets

{% code "sam/samtools-wgsim.sh" %}
{% endmd %}

{% md pileups %}
### Using Samtools: Large scale rearrangments, pileups

Use samtools to investigate genome structure. Generate pileups.

{% code "sam/samtools-pileups.sh" %}
{% endmd %}

{% endblock %}
