{% extends "unitbase.html" %}
{% load pytags %}
{# title =  RNA-Seq #}
{# subtitle = transcriptome analysis #}
{# name = RNA-Seq and transcriptome analysis  #}

{% block body %}

{% markdown %}

### Strand Specific RNA sequencing

This tutorial makes use of data from the paper

* [Comprehensive comparative analysis of strand-specific RNA sequencing methods 11(8):e1004393.
Nature Methods 7, 709â€“715 (2010) ][nmeth.1491] by Joshua Z Levin,	Moran Yassour,	Xian Adiconis,	Chad Nusbaum,
Dawn Anne Thompson, Nir Friedman,	Andreas Gnirke	& Aviv Regev.

[nmeth.1491]: http://www.nature.com/nmeth/journal/v7/n9/full/nmeth.1491.html

The tutorial demonstrates the features of stranded RNA Seq data and how to compute
counts, RPKM, and TPM values.

Follow the tutorial at {% link "strand-specific-rnaseq-intro.html" %}

- - -

### Informatics for RNA-seq: A web resource for analysis on the cloud

This tutorial is based on the paper:

* [Malachi Griffith*, Jason R. Walker, Nicholas C. Spies, Benjamin
J. Ainscough, Obi L. Griffith*. 2015.
Informatics for RNA-seq: A web resource for analysis on the cloud. 11(8):e1004393.
PLoS Computational Biology][pcbi.1004393]

[pcbi.1004393]: http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004393

Note that this tutorial uses different data analysis strategies than the publication above.
Some of the data names and annotations files have been changed and simplified. We are also pursuing
different data processing methodologies.

Follow the tutorial at {% link "griffith-intro.html" %}

- - -


{% endmarkdown %}

{% endblock %}