{% extends "unitbase.html" %}
{% load pytags %}
{# title = Quality Control #}
{# subtitle = trim, filter  #}
{# name = Quality control of sequencing data  #}

{% block body %}

{% markdown %}

The list of FASTQ quality control software is quite large, these are tools that we believe to work:

* [Trimmomatic][trimmomatic] published in Bioinformatics, 2014
* [FlexBar, Flexible barcode and adapter removal][flexbar] published in [Biology, 2012](http://www.mdpi.com/2079-7737/1/3/895)
* [CutAdapt][cutadapt] application note in [Embnet Journal, 2011](http://journal.embnet.org/index.php/embnetjournal/article/view/200)
* [Fastx Toolkit][fastx] - one of the first tools dedicated to QC
* [Skewer: a fast and accurate adapter trimmer for next-generation sequencing paired-end reads.][skewer] BMC Bioinformatics 2014.
* [PrinSeq][prinseq] application note in [Bioinformatics, 2011](http://www.ncbi.nlm.nih.gov/pubmed/21278185)
* [fastq-mcf][fastqmcf] published in [The Open Bioinformatics Journal, 2013](http://benthamscience.com/open/openaccess.php?tobioij/articles/V007/1TOBIOIJ.htm)
* [Scythe][scythe] a bayesian adaptor trimmer
* [SeqPrep][seqprep] - a tool for stripping adaptors and/or merging paired reads with overlap into single reads.
*
[shootout]: http://www.biostars.org/p/53528/
[cutadapt]: http://code.google.com/p/cutadapt/
[flexbar]: http://sourceforge.net/projects/flexbar/
[fastqmcf]: http://code.google.com/p/ea-utils/
[prinseq]: http://prinseq.sourceforge.net/
[trimmomatic]: http://www.usadellab.org/cms/index.php?page=trimmomatic
[trimgalore]: http://www.bioinformatics.babraham.ac.uk/projects/trim_galore/
[ngstoolkit]: http://www.nipgr.res.in/ngsqctoolkit.html
[fastx]: http://hannonlab.cshl.edu/fastx_toolkit/
[biopieces]: https://code.google.com/p/biopieces/
[scythe]: https://github.com/vsbuffalo/scythe
[seqprep]: https://github.com/jstjohn/SeqPrep
[tagdust]: http://genome.gsc.riken.jp/osc/english/software/
[tagcleaner]: http://tagcleaner.sourceforge.net/
[piqa]: http://www.bioinfo.uh.edu/PIQA/
[shortread]: http://www.bioconductor.org/packages/2.12/bioc/html/ShortRead.html
[seqtk]: https://github.com/lh3/seqtk
[ssuite]: http://www.bioinformatics.org/sms2/
[skewer]: https://www.biostars.org/p/104320/
[trimal]: http://trimal.cgenomics.org/

It is important to also note that not all tools implement the same functionality. Many
of these tools have specific use cases that other don't have. For example
if you wish to not trim but select data that do contain adapters then out of this list only
`cutadapt` would work etc.
{% endmarkdown %}

{% md trimmomatic %}
- - -
### Using Trimmomatic

    # Get a dataset
    fastq-dump -X 10000 --split-files SRR519926

    # Run fastqc reports on the data
    fastqc SRR519926_1.fastq

    # Trim with trimmomatic Single END mode
    trimmomatic SE -phred33 SRR519926_1.fastq output.fq SLIDINGWINDOW:4:30 MINLEN:35 TRAILING:3

    # Investigate what the result is.
    fastqc trimmomatic_1.fq

    # To perform the adapter cutting we need to find a file with adapter sequences.
    # You can create your own adapter or use the ones that come with Trimmomatic
    # Let's create an Illummina adapter file.
    echo ">illumina" > adapter.fa
    echo "AGATCGGAAGAG" >> adapter.fa

    # Let's do both quality and adapter trimming.
    trimmomatic SE SRR519926_1.fastq output.fq SLIDINGWINDOW:4:30 TRAILING:30 ILLUMINACLIP:adapter.fa:2:30:5


    # Trimmomatic in paired end mode.
    trimmomatic PE SRR519926_1.fastq SRR519926_2.fastq paired_1.fq unpaired_1.fq paired_2.fq unpaired_2.fq ILLUMINACLIP:adapter.fa:2:10:10 SLIDINGWINDOW:4:30 TRAILING:20

{% endmd %}

{% md flexbar %}
- - -
### Using Flexbar

    # Trimming with flexbar.
    flexbar -r SRR519926_1.fastq -q 30 -f sanger

    # Create a quality control plot.
    fastqc flexbar.fastq

    flexbar -r SRR519926_1.fastq -q 30 -f sanger --adapter-min-overlap 7 --adapter-trim-end RIGHT --adapters adapter.fa

    # Run the quality control plot again.
    fastqc flexbar.fastq

{% endmd %}

{% md prinseq %}
- - -
#### Using Prinseq

    #
    # Compared to Trimmomatic prinseq is about 100x slower.
    #
    # Run prinseq and trim bases from the right in a window of 5 bases
    prinseq-lite -fastq SRR519926_1.fastq -trim_qual_right 30 -trim_qual_window 4 -min_len 35 -out_good prinseq_1

    #
    # Run prinseq
    #
    fastqc prinseq_1.fq
{% endmd %}


{% endblock %}
