{% extends "unitbase.html" %}
{% load pytags %}
{# title = Genomic Variation #}
{# subtitle = SNP calling #}
{# name = Detecting simple variations: Single Nucleotide Polymorphism (SNP Calling) #}

{% block body %}

{% markdown %}
### What are SNPs?

There is a lot of imprecision in properly defining SNPs. For example the Wikipedia states:

> A single nucleotide polymorphism, also known as simple nucleotide polymorphism,
> (SNP, pronounced snip; plural snips) is a DNA sequence variation occurring commonly within a
> population (e.g. 1%) in which a single nucleotide - A, T, C or G in the genome
> (or other shared sequence) differs between members of a biological species or paired chromosomes
> [see more][wikipedia-snp]

The problem with this definition and many similar ones is that the variation that occurs may
involve more than one base. While an `A` to `C` transversion is a SNP what should
we call the `A` to `CT` change? Is that an `A` to `C` transversion plus an insertion of `T`, or
is it an insertion of `CT` in place of the `A`. Are those two SNPs or just one polymorphism?
Technically these are MNPs (Multiple Nucleotide Polymorphisms) but most may call these
as SNPs as well.

In the end even though SNP includes the words "single nucleotide" it has come to describe
all small variations (polymorphisms) in a genome. In general people use the word SNPs
to describe small nucleotide changes including insertions and deletions.

[wikipedia-snp]: https://en.wikipedia.org/wiki/Single-nucleotide_polymorphism

- - -

### Generate alignments

Even though we saw most of these commands we now add to these a few steps
that make it more realistic

    mkdir -p ~/refs/ebola

    # This will be the 2014 strain of the ebola genome.
    REF=~/refs/ebola/KJ660346.fa

    # Bam conversion shortcut.
    alias bam='samtools view -b - | samtools sort'

    # Overly lengthy chromosomal names are unwieldy.
    # Rename the sequence to just KJ660346
    efetch -db nucleotide -id KJ660346 -format fasta | seqret -filter -sid KJ660346 > $REF

    # We do need to create another index so that GATK works.
    samtools dict $REF > ~/refs/ebola/KJ660346.dict

    # Obtain the data
    fastq-dump -X 20000 --split-files SRR1972739

    # Create shortcuts to the data.
    R1=SRR1972739_1.fastq
    R2=SRR1972739_2.fastq

    # We will tag our reads by sample.
    # What this does is embeds information into the alignment
    # on the origin of the data. Allows us to have data from different
    # experiments all in the same BAM file.
    TAG='@RG\tID:mydata\tSM:ebola\tLB:library1'

    # Perform the alignment.
    bwa mem -R $TAG $REF $R1 $R2 | bam > bwa.bam

    # Index the alignment file.
    samtools index bwa.bam

- - -

### Pileups

Pileups are a data representation that show the base composition over at a given coordinate:

    # Pileup with a reference file.
    samtools mpileup -f $REF bwa.bam | head

    # KJ660346	46	G	1	^].	C
    # KJ660346	47	A	1	.	@
    # KJ660346	48	A	1	.	C
    # KJ660346	49	T	1	.	F
    # KJ660346	50	A	1	.	F

    # Pileups can take location coordinates
    samtools mpileup -f $REF bwa.bam -r KJ660346:1000-1005
    # KJ660346	1000	A	138	.$....,,.,,.,,,..,,,,,,........,,,,,,....,,,,,,,,..,,,,,,,,,,,.,,,..,,,,,,,,,,,,....,,,.....,.,,,,...,.,.........,......,.....,,,,.....^].^].^].^],	BEAEEFFEFFEHHHeEHIJJHIEnlECEEEJJJJJJFDBFJJJJJJJJFjHJJJJIIJJJJoJHJHHJHIIJJIJJJJH[JJJrJJJJJJJI9IJJJIJJJkFJIDJIJIJJGJJJJIJGJJJJGH>DEFFC<CBC@B
    # KJ660346	1001	G	138	.$.$.$.$,,.,,.,,,..,,,,,,........,,,,,,....,,,,,,,,..,,,,,,,,,,,.,,,..,,,,,,,,,,,,....,,,.....,.,,,,...,.,.........,......,.....,,,,.........,	DADDCFDFFDFFHbDHFHGFHDmlEEEEEJJJJJJFB?FJJJJJJJJFkHJIJIJGIIJJhJHJHHJHHIJJGIJJJGdJJJsJJJJIJJJ4IJIJJJJJh?IECJJIGJJHJJJJJJJJJJJGHCAEFFFDFCHCBF
    # KJ660346	1002	T	135	,,.,,.,,,..,,,,,,........,,,,,,....,,,,,,,,..,,,,,,,,,,,.,,,..,,,,,,,,,,,$,....,,,.....,,,,,...,.,.........,......,.....,,,,.........,^].^],	@BEFFDFFH`CHFHHDHEnjE@CEEJIIIJJD>@FJJJJJIJJEjHIGJHJIIIJJjIDJHEJEHIIIIIJJ;HQFCGpJJDIHIJJJJGJBGGIhFIHHIIIBJIDIIIHHGIJHJF:FACEEDFDF@?BBCCC
    # KJ660346	1003	A	135	,$,.,,.,,,..,,,,,,........,,,,,,....,,,,,,,,..,,,,,,,,,,,.,,,..,,,,,,,,,,,....,,,.....,.,,,,...,.,.........,......,.....,,,,.........,.,	@@DFDDFDD`CHHHHHHDlkDCCDDJIJIIJE@:FJJJJJHIJFfFIIJHJHIIJJoHJJHHJ@HHHHIHJIF]EHFpIJHJGJJI9IJGIFIEJgDIGGJJHGJIHHIJIHIIJIJIBDCCEFFFDFFAFDCC>
    # KJ660346	1004	C	134	,$.,,.,,,.$.,,,,,,........,,,,,,....,,,,,,,,..,,,,,,,,,,,.,,,..,,,,,,,,,,,....,,,.....,.,,,,...,.,.........,......,.....,,,,.........,.,	@DCFDFFF2CHHHFFGDmkDDDDDJIJJJIEA@EJJJJJJJJFeIJIJJJJIIJJnJIJHHJHHIJJIJJJGhGIIrJJIIIJJJ?JJIJCJIJiFIIDJJIHJJIIIJJIJJJJJIFI@CEHHFFFFFFFCCC
    # KJ660346	1005	A	135	.,,.,,,.,,,,,,........,,,,,,....,,,,,,,,..,,,,,,,,,,,.,,,..,,,,,,,,,,,....,,,.....,.,,,,...,.,.........,......,.....,,,,.........,.,^].^].^].	gCCDFFFDFHHHHHDlkDCCCDJJJJJJECCEJJJJJJIJFeIJEJJIJIJJJjJIJHHJEEIJJIJJJDjGIJsJJJJIJJJ@JJJJFiGJjFIJEJJIIJJIkJJJIJJJJJJBG@EEHGFFFFGFFCFA@<B


    # Pileups can take SAM flags in the --rf (required flag) or --ff (filtering flag)
    #
    # Pileup from reads on reverse strand.
    samtools mpileup --rf 16 -f $REF bwa.bam -r KJ660346:1,000-1,003
    #
    # KJ660346	1000	A	75	,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,^],	FFFFHHHFHIJJHIJJJJJJJJJJJJJJHJJJJIIJJJJJHHJJHIIJJIJJJJHJJJJIIJJJJFGHGH>DECB
    # KJ660346	1001	G	75	,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,	CFFFFFHCHFHGFHJJJJJJJJJJJJJJHJIJIJGIIJJJDHJJHHIJJGIJJJGJJJJJIJIJJ?HHJHCAEFF
    # KJ660346	1002	T	76	,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,$,,,,,,,,,,,,,,,,,,,,,^],	@BFFFFH@HFHHDHJIIIJJJJJJJIJJHIGJHJIIIJJICDJJEHIIIIIJJ;HJJJJJJJGJIFDGIFACEECC
    # KJ660346	1003	A	75	,$,,,,,,,$,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,	@@FDFDD@HHHHHHJIJIIJJJJJJHIJFIIJHJHIIJJHHJJJ@HHHHIHJIFIIIJIIJGIJDHHIDCCEAC>


    # A Poor man's SNP caller.
    # SNP calling with simple one-liner.
    samtools mpileup -f $REF bwa.bam | tr -d '$^],.' | awk -F '\t' 'length($5)>5 { print $0 }'

    # The poor man's snp caller can generate a BED file
    # that can be visualized in IGV.
    samtools mpileup -f $REF bwa.bam | tr -d '$^],.' | awk -F '\t' 'length($5)>5 { OFS="\t";  print $1, $2-1, $2 }'  > snps.bed

    # Building a command line alias from an awk script gets messy.
    # Both bash and awk will want to execute the variables denoted with the $ sign.
    # Best if we put the code into a separate file.
    echo 'BEGIN { FS="\t"; OFS="\t" } length($5)>5 { OFS="\t";  print $1, $2-1, $2 }' > snpcaller.awk

    alias poor_mans_snp="samtools mpileup -f $REF experiment.bam | tr -d '$^],.' | awk -f snpcaller.awk > snps.bed"

    # make that into a megaton command.
    alias megaton='generate_experimental_data; align_experimental_data; poor_mans_snp;'


* [Pileup format specification][pileup-spec] on Wikipedia.
* [5 Things to Know About SAMtools Mpileup][5things] on the MassGenomics blog.

[5things]: http://massgenomics.org/2012/03/5-things-to-know-about-samtools-mpileup.html
[pileup-spec]: https://en.wikipedia.org/wiki/Pileup_format

- - -

### The Variant Call Format

The variant call format (VCF) is a (suprisingly complex) data representation format used
to describe variations in the genome.

It is a 10+ column format where the first 9 columns are fixed the rest describe individual samples.

```
#CHROM      POS     ID     REF     ALT     QUAL    FILTER  INFO    FORMAT   SAMPLE
KM034562v1  738     .       T       C      100     PASS    AF=1.0    GT      0/1
```

* [VCF Poster][vcf-poster] is a good introductory resource on what the VCF format is.
* Short summary on VCF format: http://www.htslib.org/doc/vcf.html
* VCF Specification: http://samtools.github.io/hts-specs/

[vcf-poster]: http://vcftools.sourceforge.net/VCF-poster.pdf

- - -

### SNP calling

Samtools/BCFTools can be used to call SNPs like so:

	samtools mpileup -uvf $REF bwa.bam | bcftools call -vm -O v > bcftools.vcf

The FreeBayes SNP caller may be invoked as:

	freebayes -f $REF bwa.bam > freebayes.vcf

The Genome Analysis Tookit could be used as:

	gatk -T HaplotypeCaller -R $REF -I bwa1.bam -o gatk.vcf

### SNP calling with bcftools


- - -
### SNP calling with Freebayes

Example:

```
freebayes -f reference.fa alignment.bam > variants.vcf
```
Learn more about using `freebayes` in **{%  link "freebayes/freebayes_snpcalling.html" %}** (TODO)

- - -
### SNP calling with GATK

Example:

```
gatk -T HaplotypeCaller -R reference.fa -I alignment.bam -o variants.vcf
```

Learn more about using `gatk` in **{%  link "gatk/gatk_snpcalling.html" %}**

- - -
### How misalignment causes incorrect SNP calls.

Low complexity regions may produce counfounding alignments that are difficult to resolve.
For example a mutation with two insertions flanked by homopolymers may be very challenging to
resolve:

    CCAAACCCCCCC-T-CCCCCGCTTC
    CCAAACCCCCCCCTCCCCCCGCTTC

Align and investigate the two outputs:

    local-align.sh CCAAACCCCCCCTCCCCCGCTTC CCAAACCCCCCCCTCCCCCCGCTTC
    local-align.sh CCAAACCCCCCCTCCCCCGCTTC CCAAACCCCCCCCTCCCCCCGCTTC -gapopen 9

We will work through this problem in **{%  link "snp/snpcalling_homopolymers.html" %}**

- - -
### Additional reading

* [Validating multiple cancer variant callers and prioritization in tumor-only samples][blue-collar] on Blue Collar Bioinformatics.

[blue-collar]: https://bcbio.wordpress.com/2015/03/05/cancerval/

{%  endmarkdown %}



{% endblock %}
