{% extends "unitbase.html" %}
{% load pytags %}
{# title = SAM #}
{# subtitle = sequence alignment maps #}
{# name = Alignment representation: Sequence Alignment Maps (SAM) #}

{% block body %}

{% markdown %}

This page contains the following sections:

* [Explore the SAM Format](#format).

  You only understand the SAM format when
  you are able to describe how a change in a genome will show up
  in the SAM file.

* [Using Samtools: Basics](#samtools1)

   Converting to `BAM`, filtering alignments, using commands:
  `view`, `depth`, `tview`

- - -

Bookmark this: [Explain SAM Flags][explain-sam] Because the SAM format is so easy
to use that all we needed is a dedicated website to
decipher just a single column of it.

Good news, there is now a `flags` command to do the same:

    samtools flags PAIRED,PROPER_PAIR,REVERSE,MREVERSE
    #
    # This above lets us know that the code for it is 51 like so:
    #
    # 0x33    51    PAIRED,PROPER_PAIR,REVERSE,MREVERSE
    #

Thank you comrades!

[explain-sam]: https://broadinstitute.github.io/picard/explain-flags.html
- - -
{%  endmarkdown %}

{% md format %}
### Explore the SAM format

The Sequence Alignment Map definition: [SAM Specification][sam-spec]

[sam-spec]: https://samtools.github.io/hts-specs/

{% code "sam/explore-sam-format.sh" %}

{% endmd %}

{% md samtools1 %}

### Using Samtools: Basics

Samtools is a magical tool that alone could be used
to reproduce the majority of results in a suprising number publications.

{% code "sam/samtools-commands.sh" %}

{% endmd %}


{% endblock %}
