{% extends "unitbase.html" %}
{% load pytags %}
{# title = SNP #}
{# subtitle = SNP calling #}
{# name = Detecting simple variations: Single Nucleotide Polymorphism (SNP Calling) #}

{% block body %}

{% markdown %}
### What are SNPs?

There is a lot of imprecision in properly defining SNPs. For example the Wikipedia states:

> A single nucleotide polymorphism, also known as simple nucleotide polymorphism,
> (SNP, pronounced snip; plural snips) is a DNA sequence variation occurring commonly within a
> population (e.g. 1%) in which a single nucleotide - A, T, C or G in the genome
> (or other shared sequence) differs between members of a biological species or paired chromosomes ...
> [see more][wikipedia-snp]

The problem with this definition and many similar ones is that the variation that occurs may
involve more than one base. While an `A` to `C` transversion is a SNP what should
we call the `A` to `CT` change? Is that an `A` to `C` transversion plus an insertion of `T`, or
is it an insertion of `CT` in place of the `A`. Are those two SNPs or just one polymorphism?

In the end even though SNP includes the words "single nucleotide" it has come to describe
all small variations (polymorphisms) in a genome. Even though technically those are not SNPs.
The word INDELs are also commonly used to describe insertions/deletions in the genome.

In general SNPs designate small nucleotide changes including insertions and deletions.

[wikipedia-snp]: [https://en.wikipedia.org/wiki/Single-nucleotide_polymorphism]

- - -
### Pileups

The pileup format is a data representation that is generated from a SAM/BAM file and shows nucleotides
covering a single coordinate. The simplest SNP calling methods may use the pileup format to
identify variations.

```
chromosome   pos  base   depth    reads matches       qualities
KM034562v1  1520   G      11      ...........        12325251212
KM034562v1  1521   A      11	   C.C..CC..C.       4143.42221/
```

Samtools can generate genotype likelhoods from these coverages.

* [Pileup format specification][pileup-spec] on Wikipedia.
* [5 Things to Know About SAMtools Mpileup][5things] on the MassGenomics blog.

[5things]: http://massgenomics.org/2012/03/5-things-to-know-about-samtools-mpileup.html
[pileup-spec]: https://en.wikipedia.org/wiki/Pileup_format

Learn more about the pileup command in **{%  link "samtools_pileup.html" %}**

- - -
### The Variant Call Format

The variant call format (VCF) is a (suprisingly complex) data representation format used
to describe variations in the genome.

It is a 10+ column format where the first 9 columns are fixed the rest describe individual samples.

```
#CHROM      POS     ID     REF     ALT     QUAL    FILTER  INFO    FORMAT   SAMPLE
KM034562v1  738     .       T       C      100     PASS    AF=1.0    GT      0/1
```

* [VCF Poster][vcf-poster] is a good introductory resource on what the VCF format is.
* Short summary on VCF format: http://www.htslib.org/doc/vcf.html
* VCF Specification: http://samtools.github.io/hts-specs/

[vcf-poster]: http://vcftools.sourceforge.net/VCF-poster.pdf

- - -
### SNP calling with bcftools

`bcftools` relies on `samtools` for computing the genotype likelihoods (the probabilty for
each base to be at a given position). It then uses this likelhood to call variations:

```
samtools mpileup -uvf reference.fa alignment.bam | bcftools call -vm -O v > variants.vcf
```

Learn more about using `bcftools` in **{%  link "bcftools/bcftools_snpcalling.html" %}**

- - -
### SNP calling with Freebayes

Example:

```
freebayes -f reference.fa alignment.bam > variants.vcf
```
Learn more about using `freebayes` in **{%  link "freebayes/freebayes_snpcalling.html" %}** (TODO)

- - -
### SNP calling with GATK

Example:

```
gatk -T HaplotypeCaller -R reference.fa -I alignment.bam -o variants.vcf
```

Learn more about using `gatk` in **{%  link "gatk/gatk_snpcalling.html" %}**

- - -
### How misalignment causes incorrect SNP calls.

Low complexity regions may produce counfounding alignments that are difficult to resolve.
For example a mutation with two insertions flanked by homopolymers may be very challenging to
resolve:

    CCAAACCCCCCC-T-CCCCCGCTTC
    CCAAACCCCCCCCTCCCCCCGCTTC

Align and investigate the two outputs:

    local-align.sh CCAAACCCCCCCTCCCCCGCTTC CCAAACCCCCCCCTCCCCCCGCTTC
    local-align.sh CCAAACCCCCCCTCCCCCGCTTC CCAAACCCCCCCCTCCCCCCGCTTC -gapopen 9

We will work through this problem in **{%  link "snp/snpcalling_homopolymers.html" %}**

- - -
### Additional reading

* [Validating multiple cancer variant callers and prioritization in tumor-only samples][blue-collar] on Blue Collar Bioinformatics.

[blue-collar]: https://bcbio.wordpress.com/2015/03/05/cancerval/

{%  endmarkdown %}



{% endblock %}
