{% extends "unitbase.html" %}
{% load pytags %}
{# title = SNP calling problems #}
{# subtitle = snp calling on homopolymers #}
{# name = SNP calling for homopoymers #}

{% block body %}
{% markdown %}
Low complexity regions may produce counfounding alignments that are difficult to resolve.
For example a mutation with two insertions flanked by homopolymers may be very challenging to
resolve:

    CCAAACCCCCCC-T-CCCCCGCTTC
    CCAAACCCCCCCCTCCCCCCGCTTC

We will now attempt to recreate this situation within a realistic data analysis scenario.

{%  endmarkdown %}

{%  include_markdown "reference/download_human_data.md" %}

{%  include_markdown "reference/simulate_experiment.md" %}

{%  markdown %}

Create the modification for the `T` base at position 310:

    # Set up the shorcuts
    REF=~/refs/hs/mitoc.fa
    REAL=real.fa

    # Create the desired file
    echo ">foo" > insert.fa
    echo "CTC" >> insert.fa

    # Cut the `T` then insert the new sequence:
    cp $REF $REAL
    cutseq -filter $REF -from 310 -to 310 > temp.fa
    pasteseq -filter temp.fa insert.fa -pos 309 > $REAL

    # We want to simulate from the REAL genome and with no mutations.
    # Set the output distance between pairs to 200. Otherwise only forward
    # reads will cover the site.
    alias generate_experimental_data='dwgsim -d 200 -r 0 -N 5000 $REAL experiment'

    # Generate and create the alignment.
    generate_experimental_data; align_experimental_data

Here is what happened at position `310`

<img src="img/homopolymer-misalignment.png" class="img-responsive">

For almost all reads the reported alignment is an insertion of `CT` followed by a mismatch of
`T` to `C`:

    CCAAACCCCCCC--TCCCCCGCTTC
    CCAAACCCCCCCCTCCCCCCGCTTC

Instead of what we have actually done: we have inserted a `C` on both sides of `T`:

    CCAAACCCCCCC-T-CCCCCGCTTC
    CCAAACCCCCCCCTCCCCCCGCTTC

The reason here is that the cost of one insertion + extension + mismatch is
smaller than that of two insertions.

Let's see what our snp callers say:

    # Samtools
    samtools mpileup -uvf $REF experiment.bam | bcftools call -vm -O v > samtools.vcf

    # Freebayes
    freebayes -p 1 -f $REF --use-mapping-quality experiment.bam > freebayes.vcf

    # Run the Haplotype Caller.
    gatk -T HaplotypeCaller -R $REF -I experiment.bam -o gatk-haplotype.vcf

See {% link 'gatk/gatk_snpcalling.html' %} for the details on how to run prepare the reference for `GATK` for this
particular example.

Compare the SNP callers. Note just how wildy different their output are. Only the GATK Haplotype caller
produces the expected SNP calls.

<img src="img/homopolymer-vcf.png" class="img-responsive">

{%  endmarkdown %}

{%  endblock %}