{% extends "unitbase.html" %}
{% load pytags %}
{# title =  How to install everything #}
{# subtitle = install every piece of code in the book #}
{# name = How to install every tool in this book #}

{% block body %}
{% markdown %}

Shortcuts to installation related information:

* {% link "how-to-set-the-path" %} - to ensure that your environment finds the tools.
* {% link "bash-profile.html" %} - to ensure that your settings are applied automatically.

### How to use this page

This page lists the installation instruction for every tool in the book.
The instructions are comprehensive and can be copy pasted into a terminal.
We follow the directory layout described in {% link "unit/999-software-installation" %}

Note that for each installation step we show the update command to the `PATH`
command but we won't apply it permanently to your shell profile.
There are many ways to make use of these profiles and there is no universal way to
apply the change for everyone. See the links on top for information on where to place this command
to have it loaded automatically.

### Mac OS X

Mac OSX users need to install XCode from the AppStore. In addition users may
need also need to occasionally launch XCode manually after certain updates to agree to
changes in the license. In addition ensure that command line tools are installed with:

    xcode-select --install

The Mac OSX instructions will use [HomeBrew](http://brew.sh/) whenever the tool has a corresponding
formulat. Ocassionally when using `brew` the version of the software may older than the latest release.
In that case users should follow the Unix instructions on Mac OSX as well.
When using homebrew users need to tap the science formulas then
regularly update the formulas, especially after an XCode update.

    brew tap homebrew/science
    brew update

**Note:** even when regularly updated the brewed formulas may install
older versions of the software.

### Linux

There are numerous Linux distributions with various naming schemes. To set up a system users
need to install packages via a so called package managers: `apt-get`, `yum` and others.
Depending on the Linux distribution's intial configuration one may install some, all or none of the
package shown below. Moreover the naming of these libraries occasionally changes. Please report
errors via the ticketing system. For Ubuntu Linux the following instructions were required.

    # Install a number of required libraries.
    sudo apt-get install -y curl build-essential ncurses-dev byacc zlib1g-dev python-dev git cmake

    # Install java and python libraries.
    sudo apt-get install -y unzip default-jre python-pip

    # You may also need to install common perl modules that are not always included.
    sudo perl -MCPAN -e 'install "LWP::Simple"'
    sudo perl -MCPAN -e 'install "HTML::Entities"'

### Windows/Cygwin

TODO

- - -

## SRA toolkit

The SRA tookit allow us to access the Short Read Archive. It is a suite of tools,
the most used among them are: `prefetch`, `fastq-dump`. The SRA toolkit is
distributed as binary programs for each platform.

Webpage: http://www.ncbi.nlm.nih.gov/Traces/sra/?view=software

**Mac OSX**:

    brew update
    brew info sratoolkit
    brew install sratoolkit

**Mac OSX binary**:

    cd ~/src
    curl -OL http://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/2.5.2/sratoolkit.2.5.2-mac64.tar.gz
    tar xzvf sratoolkit.2.5.2-mac64.tar.gz

    # Update the PATH
    export PATH=~/src/sratoolkit.2.5.2-mac64.tar.gz/bin:$PATH

    # The tools that can be run are visible in
    ls ~/src/sratoolkit.2.5.2-mac64/bin/

**Unix binary**:

    cd ~/src
    curl -OL http://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/2.5.2/sratoolkit.2.5.2-ubuntu64.tar.gz
    tar xzvf sratoolkit.2.5.2-ubuntu64.tar.gz

    # Update the PATH
    export PATH=~/src/sratoolkit.2.5.2-ubuntu64/bin:$PATH

    # The tools that can be run are visible in
    ls ~/src/sratoolkit.2.5.2-ubuntu64/bin/

**Windows/Cygwin binary**:

    cd ~/src
    curl -OL http://ftp-trace.ncbi.nlm.nih.gov/sra/sdk/2.5.2/sratoolkit.2.5.2-win64.zip
    unzip sratoolkit.2.5.2-win64.zip

    # Update the PATH
    export PATH=~/src/sratoolkit.2.5.2-win64/bin:$PATH

    # The tools that can be run are visible in
    ls ~/src/sratoolkit.2.5.2-win64/bin/

- - -

## FastQC

A veritable workhorse of quality control FastQC is probably the most
commonly used software of high throughput genomics. It is a java
program that should be installed as a command line tool.

Webpage: http://www.bioinformatics.babraham.ac.uk/projects/fastqc/

**MacOS X**:

    brew update
    brew info fastqc
    brew install fastqc

**Works on all platforms that support java**:

    cd ~/src
    curl -O http://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.3.zip
    unzip fastqc_v0.11.3.zip

    # Link the fastqc executable to the ~/bin folder that
    # you have already added to the path.
    ln -sf ~/src/FastQC/fastqc ~/bin/fastqc

    # Due to what seems a packaging error
    # the executable flag on the fastqc program is not set.
    # We need to set it ourselves.
    chmod +x ~/bin/fastqc

    # The FastQC/Configuration directory contains a number of files
    # of interest that can be customized if needed.
    ls ~/src/FastQC/Configuration/

- - -

## EDirect

Entrez Direct consist of a suite of sripts written in the Perl language that can
be used to connect to the NCBI Entrez data interfaces.
The suite includes commands that coresponds to each of the Entrez API endpoints:
`esearch`, `efetch`, `elink`, `efilter`, `epost`, `einfo` as well as a tool called `xtract`
that can be used to extract elements of an XML document.

Webpage: http://www.ncbi.nlm.nih.gov/books/NBK179288/

**Mac OSX**:

    brew update
    brew info edirect
    brew install edirect

**Works on all platforms that support Perl**:

    cd ~/src
    curl ftp://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/edirect.zip -O
    unzip edirect.zip

    # This is a suite of tools thus is best to
    # add the entire directory to the search path.
    export PATH=~/src/edirect:$PATH

    # The tools that can be run are visible in source folder.
    ls ~/src/edirect

- - -

## FlexBar

Flexbar is a sequencing data quality control tool. Among
others options it offers the functionality to trim data for quality
and correct adapter contamination.

Website: http://sourceforge.net/projects/flexbar/

Cite: Matthias Dodt, Johannes T. Roehr, Rina Ahmed, Christoph Dieterich:
Flexbar âˆ’ flexible barcode and adapter processing for next-generation sequencing
platforms. MDPI Biology 2012, 1(3):895-905.

**Mac OSX**:

    cd ~/src
    curl -OL http://downloads.sourceforge.net/project/flexbar/2.5/flexbar_v2.5_macosx.tgz
    tar zxvf flexbar_v2.5_macosx.tgz

    # This program requires an extra variable to be set.
    # These can be set in the profile as well.
    export DYLD_LIBRARY_PATH=~/src/flexbar_v2.5_macosx:$DYLD_LIBRARY_PATH

    # This is a single program that can be linked to ~/bin
    ln -sf ~/src/flexbar_v2.5_macosx/flexbar ~/bin/flexbar

**Linux**:

    cd ~/src
    curl -OL http://downloads.sourceforge.net/project/flexbar/2.5/flexbar_v2.5_linux64.tgz
    tar xzvf flexbar_v2.5_linux64.tgz

    # This program requires an extra variable to be set.
    # These can be set in the profile as well.
    export LD_LIBRARY_PATH=~/src/flexbar_v2.5_macosx:$LD_LIBRARY_PATH

    # This is a single program that can be linked to ~/bin
    ln -sf ~/src/flexbar_v2.5_linux64/flexbar ~/bin/flexbar

**Windows/Cygwin**: Not available.

- - -

## Trimmomatic

Trimmomatic performs a variety of useful trimming tasks for illumina paired-end
and single ended data.  The software requires java to operate.

Website: http://www.usadellab.org/cms/?page=trimmomatic

Cite: Bolger, A. M., Lohse, M., & Usadel, B. (2014).
Trimmomatic: A flexible trimmer for Illumina Sequence Data. Bioinformatics, btu170.

**Works on all platforms that support java**

    cd ~/src
    curl -O http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/Trimmomatic-0.33.zip
    unzip Trimmomatic-0.33.zip

    # The program can be invoked via
    java -jar ~/src/Trimmomatic-0.33/trimmomatic-0.33.jar

    # To shorten the invocation above we can use command aliases or write a shell script.

    # Creating an alias.
    alias trimmomatic='java -jar ~/src/Trimmomatic-0.33/trimmomatic-0.33.jar'

    # Creating a shell script under ~/bin/trimmomatic
    # This will be a shell script that passes all command line
    # parameters to the java program with the $@ construct
    echo '#!/bin/bash' > ~/bin/trimmomatic
    echo 'java -jar ~/src/Trimmomatic-0.33/trimmomatic-0.33.jar $@' >> ~/bin/trimmomatic

    # make the shell script executable
    chmod +x ~/bin/trimmomatic

    # The ~/src/Trimmomatic-0.33/adapters/ directory contains
    # Illumina specific adapter sequences.
    ls ls ~/src/Trimmomatic-0.33/adapters/

- - -

## PrinSeq

PrinSeq is a quality control tool written in the Perl language.
It can be used to filter, reformat, or trim your genomic and metagenomic
sequence data. It generates summary statistics of your sequences in
graphical and tabular format. Note: much slower than FlexBar or Trimmomatic.

Web: http://prinseq.sourceforge.net/

Cite: Schmieder R and Edwards R: Quality control and preprocessing of metagenomic
datasets. Bioinformatics 2011, 27:863-864.

**Available on all platforms that run Perl**:

    cd ~/src
    curl -OL http://downloads.sourceforge.net/project/prinseq/standalone/prinseq-lite-0.20.4.tar.gz
    tar zxvf prinseq-lite-0.20.4.tar.gz

    # Link the program into you ~/bin folder under a shorter name.
    ln -sf ~/src/prinseq-lite-0.20.4/prinseq-lite.pl ~/bin/prinseq-lite

    # Set the executable flag on the script/
    chmod +x ~/bin/prinseq-lite


# Go into the archive and read the instructions
# located in the README file. Let's follow what it says there.
cd prinseq-lite-0.20.4





- - -

## TopHat

Note that this program requires that Bowtie2 be installed as well.

Website: https://ccb.jhu.edu/software/tophat/index.shtml

**Mac OSX** with brew:

    brew update
    brew info tophat
    brew install tophat

**Mac OSX** manual install:

    cd ~/src
    curl -O https://ccb.jhu.edu/software/tophat/downloads/tophat-2.1.0.OSX_x86_64.tar.gz
    tar xzvf tophat-2.1.0.OSX_x86_64.tar.gz

    # This is a suite of tools thus is best to
    # add the entire directory to the search path.
    export PATH=~/src/tophat-2.1.0.OSX_x86_64;$PATH'

    # See the progams included with TopHat.
    ls ~/src/tophat-2.1.0.OSX_x86_64

**Linux**:

    cd ~/src
    curl -O https://ccb.jhu.edu/software/tophat/downloads/tophat-2.1.0.Linux_x86_64.tar.gz
    tar xzvf tophat-2.1.0.Linux_x86_64.tar.gz

    # This is a suite of tools thus is best to
    # add the entire directory to the search path.
    export PATH=~/src/tophat-2.1.0.Linux_x86_64;$PATH'

    # See the progams included with TopHat.
    ls ~/src/tophat-2.1.0.Linux_x86_64

**Windows/Cygwin**:

The program is not available for Cygwin/Windows.

{% endmarkdown %}
{% endblock %}