{% extends "unitbase.html" %}
{% load pytags %}
{% load handbook %}

{# title = Introduction to RNA-Seq #}
{# subtitle = quantifying gene expression with DESeq 1 #}
{# name = Introduction to RNA-Seq #}

{% block body %}

{%  toggle "code/rnaseq1_data_install.md" id=1 title='Data Installation'  %}

{%  toggle "code/rnaseq1_shortcuts.md"  id=2 title='Shortcuts' %}

{% markdown %}

We first need to build a `hisat2` index for the genome.

	hisat2-build $REF $IDX

Running the hisat2 aligner is then as simple as:

	hisat2 $IDX -1 $R1 -2 $R2 | head

By default `histat2` produces SAM files, to convert and sort into BAM format

	hisat2 $IDX -1 $R1 -2 $R2 | sam2bam > bam/UHR_1.bam

Automate the execution with the  {% link "align-hisat2.sh" %} script that contains the following:

{%  code "code/align-hisat2.sh" %}

Execute it all:

    bash align-hisat2.sh

Visualize the results against the `hg38` human genome build for gene name `MAPK8IP2`:

{% img "img/rnaseq_intro_1.png" css="img-responsive"  %}

Run the feature counts:

    featureCounts -a $GTF -g gene_name -o gene_counts.txt bam/U*.bam  bam/H*.bam

Simplify the file by keeping only the count columns:

    cat gene_counts.txt | cut -f 1,7-12 > counts.txt

Obtain the DESeq1 code:

    curl -O https://www.biostarhandbook.com/book/rnaseq/code/deseq1.r

Pass the file through the DESeq caller:

    cat counts.txt | Rscript deseq1.r > output.txt

Our output file contains the differentially expressed genes sorted by adjusted p-values.
List the gene names and fold changes for genes that pass the 0.05% filter:

    cat output.txt | awk ' $8 < 0.05 { print $0 }' | cut -f 1,6 | head




{%  endmarkdown %}

{% endblock %}
